packages/wdio-cucumber-framework/src/utils.js

```diff
@@ -160,3 +168,96 @@ export function getStepFromFeature(feature, pickle, stepIndex, sourceLocation) {
 
     return targetStep
 }
+
+/**
+ * @param {object[]} result cucumber global result object
+ */
+export const getDataFromResult = ([{ uri }, feature, ...scenarios]) => ({ uri, feature, scenarios })
+
+/**
+ * wrap every user defined hook with function named `userHookFn`
+ * to identify later on is function a step, user hook or wdio hook.
+ * @param {object} options `Cucumber.supportCodeLibraryBuilder.options`
+ */
+export function setUserHookNames (options) {
+    const hooks = [
+        'beforeTestRunHookDefinitions',
+        'beforeTestCaseHookDefinitions',
+        'afterTestCaseHookDefinitions',
+        'afterTestRunHookDefinitions',
+    ]
+    hooks.forEach(hookName => {
+        options[hookName].forEach(testRunHookDefinition => {
+            const hookFn = testRunHookDefinition.code
+            if (!hookFn.name.startsWith('wdioHook')) {
+                const userHookAsyncFn = async (...args) => hookFn(args)
+                const userHookFn = (...args) => hookFn(args)
+                testRunHookDefinition.code = (isFunctionAsync(hookFn)) ? userHookAsyncFn : userHookFn
+            }
+        })
+    })
+}
+
+/**
+ * returns a function that calls `beforeStep` before step and then calls `afterStep` (even if step has failed)
+ * We need this workaround because `BeforeStep` and `AfterStep` hooks are not implemented in CucumberJS
+ * https://github.com/cucumber/cucumber-js/issues/997
+ * @param {Function}    code    step function
+ * @param {object}      config  wdio config
+ * @param {string}      cid     cid
+ */
+export function wrapStepWithHooks (code, config, cid) {
+    const userStepFn = async (...args) => {
+        const { uri, feature } = getDataFromResult(global.result)
+
+        // beforeStep hook
+        notifyStepHookError('BeforeStep', await executeHooksWithArgs(config.beforeStep, [uri, feature]), cid)
+
+        // step
+        let result
+        let error
+        try {
+            result = await runFnInFiberContext(code.bind(this, ...args))()
+        } catch (err) {
+            error = err
+        }
+
+        // afterStep
+        notifyStepHookError('AfterStep', await executeHooksWithArgs(config.afterStep, [uri, feature, { error, result }]), cid)
+
+        if (error) {
+            throw error
+        }
+        return result
+    }
+    return userStepFn
+}
+
+/**
+ * notify `WDIOCLInterface` about failure in hook
+ * we need to do it this way because `beforeStep` and `afterStep` are not real hooks.
+ * Otherwise hooks failure are lost.
+ * @param {string}  hookName    name of the hook
+ * @param {Array}   hookResults hook functions results array
+ * @param {string}  cid         cid
+ */
+export function notifyStepHookError (hookName, hookResults = [], cid) {
+    const result = hookResults.filter(result => result && result instanceof Error)[0]
```

Comment:
`hookResults.find(result => result instanceof Error)` should do the trick (undefined instanceof Error will simply return false and find returns the first match)